### DISCUSSION TITLE
Using shiki in Serverless Function

### DISCUSSION THREAD

**Discussion by derekr** [vercel/vercel#8726] [Help] [answered]
The shiki library internally uses `fs` to read in files located in `node_modules/shiki`. These files are understandably not copied or otherwise produced as part of the build and vercel removes any `node_module` artifacts afaict. I'm wondering what config, tools or approaches are available to allow shiki to work in a Serverless Function. The wrench:

- This is a remix project so I don't have any `api/*` functions as part of my source which negates (i think) any `functions` config in `vercel.json`.
- I'm using pnpm which I think is causing issues with things like `functions.[name].includeFiles` in `vercel.json` not work as expected.

I've tried a few things like using the `functions` config in `vercel.json` and manually copying the files, but the build output in a deployment is still opaque to me so I am literally just guessing where/how to copy these files to make them available. When instantiating a shiki highlighter I can provide the relative file path so if I can figure out where/how to copy files I think I'd mostly be unblocked.

There is also a [related discussion](https://github.com/vercel/vercel/discussions/6071) (the specific error I am seeing), but was marked as Answered prematurely.

**Comment by amyegan** [comment, marked as answer]
Hi @derekr! The related discussion says they found the answer in https://github.com/shikijs/shiki/issues/138. I think [this comment](https://github.com/shikijs/shiki/issues/138#issuecomment-1057471160) is the one with the solution (tl;dr [Node File Trace](https://github.com/vercel/nft) needs to know about the file paths that are needed for the app runtime).

If you share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), we can take a closer look to help you with debugging. It might also be worth reaching out to see if the [Shiki](https://github.com/shikijs/shiki) contributors have any advice about this since they are more familiar with the library and how it works.

**Reply by derekr** [reply]
Thanks for the quick reply! I noted that the related discussion's answer wasn't satisfactory and had tried the solution in the comment you're linking to, but it didn't work unfortunately :/.

This doesn't seem like something that will or should be fixed in the library itself, though that would be nice, but afaict the root issue is:

The library tries reading a json file in node_modules, but the file isn't accessible once deployed to vercel.

Running locally:
![image](https://user-images.githubusercontent.com/22249/195687476-4c3e1dda-f0be-4ec6-9ac1-27b5e278c70f.png)


Baseline example:
This is without the fix mentioned in the related issue thread.
https://remix-shiki-vercel.vercel.app
https://github.com/derekr/remix-shiki-vercel/commit/bdbbf9d88ead64d9003f2a1ce3691d83e5bc8e30
![image](https://user-images.githubusercontent.com/22249/195687328-3b73a187-6d01-4656-ac28-187967dafcf0.png)

Suggested fix example:
https://remix-shiki-vercel-drk.vercel.app
https://github.com/derekr/remix-shiki-vercel/pull/1
![image](https://user-images.githubusercontent.com/22249/195687238-88a718e0-4013-4f24-8bde-3ac0ea7ce198.png)

**Reply by lili21** [reply]
same issue here
